### DISCUSSION TITLE
What's the design checkpoint in the example bitstream manipulation code?

### DISCUSSION THREAD

**Discussion by yu-zou** [Xilinx/RapidWright#452] [Q&A] [answered]
In the example bitstream manipulation code, provided in the document. There are two inputs, one is checkpoint, and the other is bitstream.

My questions are:
1. Is the checkpoint argument required?
2. If true, then what kind of dcp can be used as an input, e.g. a dcp post implementation is required?

Thank you, Yu.
```java
public static void main(String[] args) {
  Design design = Design.readCheckpoint(args[0]);
  Bitstream bitstream = Bitstream.readBitstream(args[1]);
  ConfigArray configArray = bitstream.configureArray();

  // Changes the initialization of the FF to 1
  Cell cell = design.getCell("myFF");
  cell.getProperty("INIT").setValue("1");
  configArray.updateUserStateBits(cell);
  bitstream.updatePacketsFromConfigArray();

  design.writeCheckpoint(args[2]);
  bitstream.writeBitstream(args[3]);
}
```

**Comment by clavin-xlnx** [comment, marked as answer]
Hi Yu,

To answer your questions, yes, RapidWright will need a placed and routed DCP as input in order to know the location of the cell being updated.  You'll want the same DCP used to create the original bitstream.  These APIs are meant to make incremental changes to an existing bitstream.  It can't create a functional bitstream from scratch.

Chris
